src/Chainweb/Pact/Backend/PactState/GrandHash.hs

```diff
@@ -0,0 +1,705 @@
+      ++
+      [ "    ]"
+      , ")"
+      ]
+
+    makeChainMap :: HashMap ChainId (ByteString, BlockHeader) -> [String]
+    makeChainMap = map (uncurry makeChainEntry) . List.sortOn fst . HM.toList
+
+    makeChainEntry :: ChainId -> (ByteString, BlockHeader) -> String
+    makeChainEntry cid (hash, header) =
+      let
+        jsonDecode j = "unsafeJsonDecode @BlockHeader " ++ j
+        fromHex b = "unsafeFromHex " ++ b
+        sCid = Text.unpack (chainIdToText cid)
+        sHash = inQuotes $ Text.unpack (hex hash)
+        sHeader = embedQuotes $ Text.unpack (J.encodeText (J.encodeWithAeson header))
+      in
+      concat
+        [ "(unsafeChainId " ++ sCid ++ ", "
+        , "EncodedSnapshot (" ++ fromHex sHash ++ ") (" ++ jsonDecode sHeader ++ ")"
+        , ")"
+        ]
+
+    prefix = List.unlines
+      [ "-- NOTE: This module has been auto-generated."
+      , "-- Do not edit it."
+      , ""
+      , "{-# LANGUAGE ImportQualifiedPost #-}"
+      , "{-# LANGUAGE OverloadedStrings #-}"
+      , "{-# LANGUAGE TypeApplications #-}"
+      , ""
+      , "module Chainweb.Pact.Backend.PactState.EmbeddedHashes"
+      , "  ( EncodedSnapshot(..)"
+      , "  , grands"
+      , "  )"
+      , "  where"
+      , ""
+      , "import Chainweb.BlockHeader (BlockHeader)"
+      , "import Chainweb.BlockHeight (BlockHeight(..))"
+      , "import Chainweb.ChainId (ChainId, unsafeChainId)"
+      , "import Data.Aeson qualified as A"
+      , "import Data.ByteString (ByteString)"
+      , "import Data.ByteString.Base16 qualified as Base16"
+      , "import Data.List qualified as List"
+      , "import Data.Map (Map)"
+      , "import Data.Map qualified as Map"
+      , "import Data.Ord (Down(..))"
+      , "import Data.Text (Text)"
+      , "import Data.Text.Encoding qualified as Text"
+      , ""
+      , "data EncodedSnapshot = EncodedSnapshot"
+      , "  { pactHash :: ByteString"
+      , "  , blockHeader :: BlockHeader"
+      , "  }"
+      , ""
+      , "unsafeJsonDecode :: (A.FromJSON a) => Text -> a"
+      , "unsafeJsonDecode t = case A.decodeStrict (Text.encodeUtf8 t) of"
+      , "  Just a -> a"
+      , "  Nothing -> error \"EmbeddedHashes: invalid json construction\""
+      , ""
+      , "unsafeFromHex :: Text -> ByteString"
+      , "unsafeFromHex t = case Base16.decode (Text.encodeUtf8 t) of"
+      , "  Right a -> a"
+      , "  Left err -> error $ \"EmbeddedHashes: unsafeFromHex failed: \" ++ show err"
+      , ""
+      , "-- | sorted in descending order."
+      , "grands :: [(BlockHeight, Map ChainId EncodedSnapshot)]"
+      , "grands = List.sortOn (Down . fst)"
+      , "  ["
+      , unlines (makeEntries input)
+      , "  ]"
+      ]
+
+exitLog :: (Logger logger) => logger -> Text -> IO a
+exitLog logger msg = do
+  logFunctionText logger Error msg
+  exitFailure
+
+hashMapToMap :: (Hashable k, Ord k) => HashMap k a -> Map k a
+hashMapToMap = Map.fromList . HM.toList
+
+withConnections :: (Logger logger)
```

Comment:
This is not exception safe, but is useful in this module. That should be documented.